zarith/zarith.go

```diff
@@ -128,15 +208,91 @@ func Encode(value *big.Int) ([]byte, error) {
 	return bitStringToBytes(encodedBitString), nil
 }
 
-// EncodeToHex encodes a number to zarith
-func EncodeToHex(value *big.Int) (string, error) {
-	bytes, err := Encode(value)
+// EncodeSignedToHex encodes an unsigned integer to zarith
+func EncodeSignedToHex(value *big.Int) (string, error) {
+	bytes, err := EncodeSigned(value)
 	if err != nil {
 		return "", err
 	}
 	return hex.EncodeToString(bytes), nil
 }
 
+// DecodeSigned decodes a zarith encoded signed integer from the entire input byte array.
+// Assumes the input contains no extra trailing bytes.
+func DecodeSigned(source []byte) (*big.Int, error) {
+	if len(source) == 0 {
+		return nil, xerrors.New("expected non-empty byte array")
+	}
+
+	// Split input into 8-bit bitstrings
+	segments := make([]string, len(source))
+	for i, curByte := range source {
+		segments[i] = fmt.Sprintf("%08b", curByte)
+	}
+
+	// Trim off leading continuation bit from each segment
+	for i, segment := range segments {
+		segments[i] = segment[1:]
+	}
+
+	// Trim off the sign flag from the first segment
+	firstSegment := []rune(segments[0])
+	isNegative := firstSegment[0] == '1'
+	segments[0] = string(firstSegment[1:])
+
+	// Reverse the order of the segments.
+	// Source: https://github.com/golang/go/wiki/SliceTricks#reversing
+	for i := len(segments)/2 - 1; i >= 0; i-- {
```

Comment:
I prefer counting up, but it should work either way. IMO the two indices version is easier to read btw